app/assets/javascripts/components/generic_object/generic-object-definition-treeview-component.js

```diff
@@ -0,0 +1,168 @@
+ManageIQ.angular.app.component('genericObjectDefinitionTreeviewComponent', {
+  bindings: {
+    recordId: '@?',
+    showListUrl: '@',
+    showUrl: '@',
+  },
+  controllerAs: 'vm',
+  controller: genericObjectDefinitionTreeviewController,
+  templateUrl: '/static/generic_object/generic_object_definition_treeview.html.haml',
+});
+
+genericObjectDefinitionTreeviewController.$inject = ['API', 'miqService', '$window', '$timeout'];
+
+function genericObjectDefinitionTreeviewController(API, miqService, $window, $timeout) {
+  var vm = this;
+
+  vm.treeData = [];
+  vm.activeNode = {};
+
+  vm.$onInit = function() {
+    $timeout(function() {
+      if ($window.top.name !== '') {
+        vm.treeData = $window.top.name;
+        $window.top.name = '';
+      } else {
+        setGenericObjectDefinitionNodes();
+      }
+      vm.setActiveNode();
+    });
+  };
+
+  vm.nodeSelect = function(node) {
+    var key = node.key.split('_');
+
+    miqService.sparkleOn();
+    switch (key[0]) {
+      case 'god':
+        if (key[1] === 'root') {
+          $window.location.href = vm.showListUrl;
+        } else {
+          $window.location.href = vm.showUrl + '/' + key[1];
+        }
+        break;
+      case 'cbs':
+        $window.location.href = vm.showUrl + '/' + node.parentNodeId + '?cbs=' + key[1];
```

Comment:
So.. you can actually convince the browser to change the URL without reloading the whole page...

docs: https://developer.mozilla.org/en-US/docs/Web/API/History_API

example:
`history.pushState({}, "new_title", new_url);`

(you will probably want to do `miqAjax` to replace the partials together with this)